src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOperation.scala

```diff
@@ -3,64 +3,60 @@ package core.instance.update
 
 import mesosphere.marathon.core.condition.Condition
 import mesosphere.marathon.core.instance.{Goal, Instance}
-import mesosphere.marathon.core.instance.Instance.AgentInfo
-import mesosphere.marathon.core.task.{Task, TaskCondition}
+import mesosphere.marathon.core.instance.Instance
+import mesosphere.marathon.core.task.TaskCondition
 import mesosphere.marathon.state.Timestamp
 import org.apache.mesos
 
-import scala.collection.immutable.Seq
-
 sealed trait InstanceUpdateOperation {
   def instanceId: Instance.Id
-  /**
-    * The possible task state if processing the state op succeeds. If processing the
-    * state op fails, this state will never be persisted, so be cautious when using it.
-    */
-  def possibleNewState: Option[Instance] = None
 }
 
 object InstanceUpdateOperation {
   /** Launch (aka create) an ephemeral task*/
   case class LaunchEphemeral(instance: Instance) extends InstanceUpdateOperation {
     override def instanceId: Instance.Id = instance.instanceId
-    override def possibleNewState: Option[Instance] = Some(instance)
   }
 
   /** Revert a task to the given state. Used in case TaskOps are rejected. */
   case class Revert(instance: Instance) extends InstanceUpdateOperation {
     override def instanceId: Instance.Id = instance.instanceId
-    override def possibleNewState: Option[Instance] = Some(instance)
+  }
+
+  /**
+    * * Reschedules resident instance that already has a reservation but became terminal (reserved).
+    *
+    * @param reservedInstance already existing reserved instance that is now in scheduled state.
+    */
+  case class RescheduleReserved(reservedInstance: Instance, runSpecVersion: Timestamp) extends InstanceUpdateOperation {
```

Comment:
Is this here because `Reserve` is usually used to launch resident tasks?

Reply:
@alenkacz, do you have details? I've forgot the reason.

Reply:
we launch residents in two steps - we first Reserve (create the reservation, don't launch anything) and when we get offer back with that reservation, we launch

This operation is here for instances, that have reservation and were launched before

Reserve is for fresh instances, that haven't been launch before and DON't have reservation